job_creator.php

```diff
@@ -0,0 +1,273 @@
+        }
+        // phpcoverage also runs unit tests
+        // always run on silverstripe account
+        if ($run['phpcoverage'] || preg_match('#^silverstripe/#', $githubRepository)) {
+            if ($simpleMatrix) {
+                $matrix['include'][] = $this->createJob(0, [
+                    'phpcoverage' => true
+                ]);
+            } else {
+                $matrix['include'][] = $this->createJob(2, [
+                    'db' => DB_MYSQL_57_PDO,
+                    'phpcoverage' => true
+                ]);
+            }
+        }
+        // endtoend / behat
+        if ($run['endtoend'] && file_exists('behat.yml')) {
+            $matrix['include'][] = $this->createJob(0, [
+                'endtoend' => true,
+                'endtoend_suite' => 'root'
+            ]);
+            if (!$simpleMatrix) {
+                $matrix['include'][] = $this->createJob(3, [
+                    'db' => DB_MYSQL_80,
+                    'endtoend' => true,
+                    'endtoend_suite' => 'root'
+                ]);
+            }
+        }
+        // javascript tests
+        if ($run['js'] && file_exists('package.json')) {
+            $matrix['include'][] = $this->createJob(0, [
+                'js' => true
+            ]);
+        }
+        return $matrix;
+    }
+
+    public function createJson(array $inputs): string
+    {
+        // $myRef will either be a branch for push (i.e cron) and pull-request (target branch), or a semver tag
+        $myRef = $inputs['github_my_ref'];
+        $isTag = preg_match('#^[0-9]+\.[0-9]+\.[0-9]+$#', $myRef, $m);
+        $branch = $isTag ? sprintf('%d.%d', $m[1], $m[2]) : $myRef;
+
+        $githubRepository = $inputs['github_repository'];
+        $this->installerVersion = $this->getInstallerVersion($githubRepository, $branch);
+
+        $run = [];
+        $extraJobs = [];
+        $dynamicMatrix = true;
+        $simpleMatrix = false;
+        foreach ($inputs as $input => $value) {
+            if (in_array($input, ['endtoend', 'js', 'phpunit', 'phpcoverage', 'phplinting'])) {
+                $run[$input] = $this->parseBoolValue($value);
+            } else if ($input === 'extra_jobs') {
+                if ($value === 'none') {
+                    $value = [];
+                }
+                $extraJobs = $value;
+            } else if ($input === 'dynamic_matrix') {
+                $dynamicMatrix = $this->parseBoolValue($value);
+            } else if ($input === 'simple_matrix') {
+                $simpleMatrix = $this->parseBoolValue($value);
+            } else if (in_array($input, ['github_my_ref', 'github_repository'])) {
+                continue;
+            } else {
+                throw new LogicException("Unhandled input $input");
+            }
+        }
+        $matrix = ['include' => []];
+
+        if ($dynamicMatrix) {
+            $matrix = $this->buildDynamicMatrix($matrix, $run, $simpleMatrix, $githubRepository);
+        }
```

Comment:
```suggestion
        if ($dynamicMatrix) {
            $matrix = $this->buildDynamicMatrix($matrix, $run, $simpleMatrix, $githubRepository);
        } else {
            $matrix = ['include' => []];
        }
```
Removes the temptation of adding more matrix stuff before `if ($dynamicMatrix) {` which would result in stuff being added that then gets immediately overwritten by the `$matrix = $this->buildDynamicMatrix(` assignment.